### DISCUSSION TITLE
Cannot access artifacts in folders in DNG with ELM Python Client

### DISCUSSION THREAD

**Discussion by TheGreatConfundus** [IBM/ELM-Python-Client#44] [Q&A] [answered]
Hi,

Some time ago I asked in the Jazz forum about the use of APIs in DNG . They indicated me that the Reportable API and the OSLC API were the way to go, as the REST API is private and undocumented.

Right now we are using the Reportable API and applying filtering in the JavaScript code. We are able to run the extensions in DNG via the XML gadget , still our problem is to retrieve the data, load the JavaScript extension and update the front custom metrics based in the filtered data. Our problem is that, whenever the extension loads, it needs to retrieve all the data of project, make the proper filtering, taking a frustrating amount of time to load.

After a lot of research in the Jazz.net forum, I found a comment mentioning this GitHub Python application. We still do not know how we could implement it to the JavaScript project extensions we currently have, but it might be a better way to filter and retrieve the data. I have tried to use the "oslc_query" script, but I am unable to access the artifacts of a view (or inside a folder) of a specific project area. The log seems to indicate that the connection is effectively ocurring.

oslcquery -J https://abc.clm.com -U -P  -p "P0M10121" -F "P0M10121 - Initial Stream" -q "rm_nav:parent=$""01 System Breakdown Structure (SBS)""" -s dcterms:title -O test1.csv

oslcquery -J https://abc.clm.com -U -P  -p "P0M10121" -F "P0M10121 - Initial Stream" -r rm_view:"Prueba1" -s dcterms:title -O test2.csv

What am I doing wrong? There seems to be some sort of error regarding the string and the spaces after the "-q" and "-r" commands (I see there was another thread on the same topic, but the provided solution is not clear to me). 

Thanks for the development of the application, and also for your answer in advance.

Pablo

**Comment by barny** [comment]
> the REST API is private and undocumented

Not sure which "REST API" you're referring to, so can't comment.

For me, your command #1 works once adapted to my server's URL/user/password/project/configuration/foldername without modifying the -q value  - but obviously the folder "01 System Breakdown Structure (SBS)" (case- and space-sensitive) must exist - if it doesn't you get a rather long/messy error message including text like:

`Folder name 01 System Breakdown Structure (SBS) not found`

then

`Exception: Error parsing query`

Your command #2 will never work because the oslcquery -r option doesn't accept anything like a view name. The OSLC Query API doesn't know anything about views. Also "rm_view:"Prueba1"" isn't valid syntax for an OSLC Query oslc.where query. Add -h to the command to see all the options.

Trying to decode what you say it sounds like you're trying to write a DNG Extension widget (you refer to Javascript) which you want to do some analysis of artifacts in a view and (I guess) present some information to the user. Doesn't the extensions API allow retrieving artifacts in the current view? https://jazz.net/wiki/bin/view/Main/RMExtensions702

I thought examples like the ASIL guidance widget process visible artifacts to update an ASIL attribute, see https://jazz.net/wiki/bin/view/Main/RMExtensionsExamples702

I'm no expert on DNG extensions. I doubt you're going to be able to invoke `oslcquery` from a DNG extension. You may be able to program a query for the OSLC Query API from an extension but I've no example of doing that.

**Comment by barny** [comment, marked as answer]
If you want to use OSLC Query to get all artifacts in a module, use e.g. (on Windows)

`oslcquery -J https://jazz.ibm.com:9443 -U ibm -P ibm -p "rm_optin_p1" -F "rm_optin_p1 Initial Stream" -s * -q "rm:module=^""AMR Stakeholder Requirements Specification"""`

This uses my enhanced query notation - you use ^ followed by the name of a module and will find its URI to put into the OSLC Query that's executed. There's a a similar notation $"01 Requirements" to get the URI of a folder, and also ~1234 will find the URI of an artifact with identifier 1234.

Obviously you'll have to adapt to your server/project/stream/module name.

Add -V if you want to see the query url that's generated, in this case:

`https://jazz.ibm.com:9443/rm/views?oslc.query=true&projectURL=https%3A//jazz.ibm.com%3A9443/rm/process/project-areas/_tg0CsFEhEe2d5dG-54qhbg&oslc.prefix=rm%3D%3Chttp%3A//www.ibm.com/xmlns/rdm/rdf/%3E&oslc.where=rm%3Amodule%3D%3Chttps%3A//jazz.ibm.com%3A9443/rm/resources/MD_tuvxFlEhEe2d5dG-54qhbg%3E&oslc.select=%2A&oslc.paging=true&oslc.pageSize=200`

Note the quoting - Windows needs the whole -q value surrounded by quotes because it contains a space, and inside the quotes existing quotes have to be doubled.

**Reply by TheGreatConfundus** [reply]
Hi again,

Thanks for your answers. I finally found the way to make the query of the folder, in my case single quotes around the "-q" parameter needs to be used:
_**oslcquery -J  -U  -P  -p "P0M10146 " -F "P0M10146 - Initial Stream" -s dcterms:identifier,dcterms:title -q 'rm_nav:folder=$""01 System Breakdown Structure""' --saveprocessedresults resultados/rmresults_con_title -V

However, we still have two problems:

1) In some projects we have a large amount of data. When we try to query, it seems that there is a limit of 15000 results, is there any way to surpass this threshold?
![image](https://github.com/IBM/ELM-Python-Client/assets/132047094/c68bd43b-9b10-41be-8e11-5d168e865515)

2) What we really would like to get is the URL that returns the filtered data outside this app in JSON format, e.g.: on a new tab (instead of saving it to a file). Right now it seems like the requested URL is too large to be loaded.
![image](https://github.com/IBM/ELM-Python-Client/assets/132047094/e0040f89-6ad9-4482-b720-7367ea42a3b6)
![image](https://github.com/IBM/ELM-Python-Client/assets/132047094/4b1838bb-3c94-40b1-bb9d-0f7adec629a4)

And when we try to access to the shorter URL given by the "-V" command, the following web error appears:
![image](https://github.com/IBM/ELM-Python-Client/assets/132047094/947fc400-f273-4ce9-ae2b-18b382a2c625)


To put you in more context, we are developing an "OpenSocial gadget" to visualize custom metrics on the dashboard of the "IBM Engineering Requirements Management DOORS Next (/rm)" module. Our problem arises when we try to retrieve the data from DNG. Right now we are retrieving the data in the following way: 
**_<domain>/rm/publish/resources?projectName=<projectName> + filters._** 

We are not using the RM API that you mention as it is very intrusive and severely interfers with DNG performance. Currently, we are adding some more filters from the [Reportable Rest API documentation (https://jazz.net/wiki/bin/view/Main/RRCReportableRestAPI). . However, these filters fall short to us; the responses consist of 100,000 lines and take 40-60 seconds. We need to add more filters and perform more specific queries that this API (/rm/publish/) cannot handle. That is why we think your OSLC-query based tool can be of great help to us, as it can create complex queries for targeting specific attributes of specific folders in DNG.

Thanks in advance,

Greets.

Pablo

**Reply by barny** [reply]
> a limit of 15000 results

There's no built in limit in my code, and AFAIK also not in DOORS Next. The line you show indicates there are 97270 results so the server knows how much it has to give you. What's happening on the server when it pauses at 15000?

There is a -M nnnn option to stop requesting pages after this nnnn results have been received.

I've used oslcquery on a large test server without issues. Can't remember the details but I'm sure it had more that 15k artifacts in the component/project

oslcquery uses caching to speed up repeated GET operations (note queries don't get cached) which can sometimes break stuff when the cached data disagrees with some other responses - add option -W to clear the cache and then continue with caching, or -WW to completely disable caching.

> the URL that returns the filtered data outside this app in JSON format

OSLC Query uses rdf. Also results are by paged default to 200 at a time so there isn't a single url to get the results.

You can try disabling paging but then your poor server has to put all the data into one response. To disable paging add option --pagesize 0

> URI too long

That URL looks very odd because there are at least three oslc_config.context parameters, all identical, and it looks like you cut more off the end. Only one of these is needed, there shouldn't be repeats.

How did you generate that URL? Is it for the first page of results?

**Reply by barny** [reply]
looks like there may be an issue as I'm seeing ever-longer URLs for paged results.

**Reply by barny** [reply]
My code provides both a oslc_config.context parameter and a Configuration-Context header which is over-specifying but I wasn't aware of an issue - If I remove the header the next page URLs don't keep extending.

If you can edit oslcqueryapi.py, there's a line around 548
`        headers = {}
`
Change this to suppress the Configuration-Context header:
`        headers = {'Configuration-Context': None}
`
Let me know if this helps.

**Reply by barny** [reply]
You should probably increase the page size (1000, 5000?) to speed up collecting this large number of results, but keep an eye on server load because such large queries are likely to impact other users of the server.

**Reply by TheGreatConfundus** [reply]
Hello again,

Thanks one more time for your answers, they are of great help. We finally could get the full data of a project thanks to your last comment on increasing the page size:). 

We have one more question. We realized that, in the OSLC query response, there are a linked data that is missing. I attach a screenshot comparing the data of an artifact requirement that we recover using the Reportable API (left) and using OSLC (right):

![image](https://github.com/IBM/ELM-Python-Client/assets/132047094/00115c45-59bf-419a-ad12-6d9cabc24772)

Is there any way to obtain the data from the traceability links as in the Reportable API? I mean, not by recovering the link data it into another query, but to configure the URL query in such a way that we can get another's linked data information as in the example above.

Our goal is to use the OSLC to filter by data attributes with conditions of the kind: -q "( Status=Approved ) ||  ( ( Status=Rejected and Priority=Low ) && ( Severity=High) )", that currently we are not able to do in the Reportable API. On this manner we could get more filtered, faster queries so that our extensions do not take so long to load.

Thanks,

Pablo

**Reply by barny** [reply]
:-)

With OSLC Query you can only get outgoing links from the artifact, i.e. RM->RM links as all other links are incoming to RM and aren't stored on the RM artifacts. I've definitely received links for the Satisfies link, it has to be included in the oslc.select (option `-s Satisfies` in oslcquery). Have you tried with oslc.select=* (-s *)? I just checked and e.g. `Child of` is returned (the link is stored at the child end so is only returned on that artifact).

**Reply by barny** [reply]
I can get the child of link using `-s "'Child Of'"`

**Reply by barny** [reply]
re: your query ( Status=Approved ) || ( ( Status=Rejected and Priority=Low ) && ( Severity=High) )

This requires three OSLC queries, and I think it can be simplified to:

( Status=Approved ) || ( Status=Rejected and Priority=Low and Severity=High )

which requires two queries
